### PR TITLE
better support of ipv6-only clusters: prevent creation of default-ipv4-pool during calico-node startup

### PR DESCRIPTION
**Pull request by kruftik**
## Description

Current calico-node startup logic supports creation of default ip-pools both for ipv4 and ipv6 stacks if the ones do not exist. However, it does not take into account accordingly the ipv6-only clusters (without ipv4 addresses on any interfaces at all): as soon as default-ipv4 pool being created, every calico-node instance begin log flooding with something like: 

```
felix/route_table.go 1016: Failed to get link attributes error=interface not present ifaceRegex="^vxlan.calico$" ipVersion=0x4 tableIndex=254
````

This PR mitigate such an issue: in addition to `NO_DEFAULT_POOLS` environment variable, `CALICO_IPV4POOL_CIDR`=`none` and `CALICO_IPV6POOL_CIDR`=`none` special values are added.

## Release Note

```release-note
In order to prevent default IP-pools creation, `CALICO_IPV4POOL_CIDR`=`none` and `CALICO_IPV6POOL_CIDR`=`none` environment variable special values are now supported.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Comment by caseydavenport** [comment]
I wonder if instead of introducing new env vars for this, we should follow the pattern we used for `CALICO_NETWORKING_BACKEND` and support a `none` option for the existing `IPV4POOL_CIDR` variable.

That said, I am curious why using `NO_DEFAULT_POOLS` and then creating an IPv6 pool out-of-band isn't viable here? That would be my preferred solution here, since in general I would like us to move away from using calico/node to create IP pools - configuring APIs from a collection of env vars makes it hard to support all of the necessary use cases. For now though, calico/node is where default IP pools are created, so I'm not completely against merging something like this as a stopgap.

**Comment by kruftik** [comment]
@caseydavenport ,

> we should follow the pattern we used for `CALICO_NETWORKING_BACKEND` and support a `none` option for the existing `IPV4POOL_CIDR` variable.

yeah, it's a better option. I will refactor the PR into such an approach.

**Comment by kruftik** [comment]
@caseydavenport , done. Could you look into?

File: node/pkg/lifecycle/startup/startup.go [hunk 1]
```diff
@@ -762,16 +762,33 @@ func configureIPPools(ctx context.Context, client client.Interface, kubeadmConfi
...
+	)
+
+	if ipv4Pool == "none" {
+		log.Info("Skipping IPv4 pool configuration")
+
+		ipv4PoolEnabled = false
+	}
+
+	if ipv6Pool == "none" {
+		log.Info("Skipping IPv4 pool configuration")
```
Comment (com30n) [inline node/pkg/lifecycle/startup/startup.go:777]: It looks like it should be `Skipping IPv6 pool configuration`

**Review by caseydavenport — CHANGES_REQUESTED** [review]
Apologies for the delay here - this looks good to me!

I do think we need a pair of unit tests to verify this behavior works - here's an example of a similar test you can base it on: https://github.com/projectcalico/calico/blob/8d31b2669e8cde1558b128fc7db3015970921e90/node/pkg/lifecycle/startup/startup_test.go#L455-L477

**Comment by caseydavenport** [comment]
@kruftik is this still on your radar? The only thing remaining is a unit test (see my last comment)

**Comment by kruftik** [comment]
> @kruftik is this still on your radar? The only thing remaining is a unit test (see my last comment)

seems like done

**Review by caseydavenport — APPROVED** [review]

**Pull request merged**